flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java

```diff
@@ -162,20 +168,33 @@ public List<String> listDatabases() throws CatalogException {
         .collect(Collectors.toList());
   }
 
+  private void registerPartitionFunction(String databaseName) {
+    functions.putIfAbsent(
+        new ObjectPath(databaseName, PartitionTransformUdf.Bucket.FUNCTION_NAME),
+        new CatalogFunctionImpl(
+            PartitionTransformUdf.Bucket.class.getName(), FunctionLanguage.JAVA));
+    functions.putIfAbsent(
+        new ObjectPath(databaseName, PartitionTransformUdf.Truncate.FUNCTION_NAME),
+        new CatalogFunctionImpl(
+            PartitionTransformUdf.Truncate.class.getName(), FunctionLanguage.JAVA));
+  }
+
   @Override
   public CatalogDatabase getDatabase(String databaseName)
       throws DatabaseNotExistException, CatalogException {
     if (asNamespaceCatalog == null) {
       if (!getDefaultDatabase().equals(databaseName)) {
         throw new DatabaseNotExistException(getName(), databaseName);
       } else {
+        registerPartitionFunction(databaseName);
         return new CatalogDatabaseImpl(Maps.newHashMap(), "");
       }
     } else {
       try {
         Map<String, String> metadata =
             Maps.newHashMap(asNamespaceCatalog.loadNamespaceMetadata(toNamespace(databaseName)));
         String comment = metadata.remove("comment");
+        registerPartitionFunction(databaseName);
```

Comment:
@wuwenchi, these functions should not be registered in every known database. The problem before was not how you were tracking functions, it was the behavior. You need to change the behavior.

Here's my last comment:

> The problem is that this is effectively registering the functions in every database. Instead, choose one database where you want the functions to live. For stored procedures, we use `system`. I think that's a good idea here.

Please remove the registration code that isn't needed and update the function lookup and listing to only return these functions if the database is `system`.

Reply:
About the `system` database, is this database just a logical database? Is it that users cannot create tables under this library? In sql, is it not possible to execute `use iceberg_catalog.system`?  Because this is related to the list of functions, if we you cannot use `use system` to switch the currently used database, we cannot list functions under the `system`, because `show functions` can only display the functions under the current database.

Reply:
I don't see why `system` would be different than any other database. It should be a normal database.

Reply:
If it is a normal database, should we create it first when open catalog  like the `default` database? Because if it is not created, the user cannot see the database, but can use the functions under the `system` database, which feels a little strange.